docs/design-documents/drivers/serial/serial.md

```diff
@@ -0,0 +1,317 @@
+    *  @param tx_buffer The transmit buffer 
+    *  @param tx_buffer_size The size of the transmit buffer 
+    */
+    BufferedSerial(PinName tx, 
+                   PinName rx, 
+                   int baud = MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE, 
+                   void *rx_buffer,
+                   size_t rx_buffer_size,
+                   void *tx_buffer,
+                   size_t tx_buffer_size);
+```
+
+ The rest of the class API is that of `UARTSerial` and is described [here](https://os.mbed.com/docs/mbed-os/v5.14/mbed-os-api-doxy/classmbed_1_1_u_a_r_t_serial.html).
+
+
+### Detailed design : UnbufferedSerial
+
+`UnbufferedSerial` is a new class that adds a `FileHandle` interface to `RawSerial`. `RawSerial` is the preferred implementation for unbuffered I/O but it lacks a `FileHandle` interface and so it cannot be used by the retarget code. Instead the retarget code uses a non-public `DirectSerial` class which adds to the complexity of classes and should be removed. `UnbufferedSerial` will not provide any of the printf methods that `RawSerial` has; applications should use _fdopen_ to get a _FILE*_ to use the printf function, see [Usage scenarios and examples](#usage-scenarios-and-examples).
+
+The below diagram shows the detailed inheritance hierarchy for `UnbufferedSerial`.
+
+![UnbufferedSerial class](./diagrams/unbuffered-serial.png)
+
+**API description**
+
+This is the proposed API for the `UnbufferedSerial` class.
+
+```C
+class UnbufferedSerial: public SerialBase, public FileHandle, private NonCopyable<UnbufferedSerial> {
+
+public:
+    /** Create a UnbufferedSerial port, connected to the specified transmit and receive pins, with the specified baud.
+     *
+     *  @param tx Transmit pin
+     *  @param rx Receive pin
+     *  @param baud The baud rate of the serial port (optional, defaults to MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE)
+     *
+     *  @note
+     *    Either tx or rx may be specified as NC if unused
+     */
+    UnbufferedSerial(PinName tx, PinName rx, int baud = MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE);
+
+    virtual ssize_t write(const void *buffer, size_t size);
+    virtual ssize_t read(void *buffer, size_t size);
+    virtual off_t seek(off_t offset, int whence = SEEK_SET);
+    virtual off_t size();
+    virtual int isatty();
+    virtual int close();
+    virtual short poll(short events) const;
+
+#if !(DOXYGEN_ONLY)
+protected:
+
+    /* Acquire exclusive access to this serial port
+     */
+    virtual void lock(void);
+
+    /* Release exclusive access to this serial port
+     */
+    virtual void unlock(void);
+#endif
+};
+```
+
+# Usage scenarios and examples
+
+### Scenario 1 `Printing to stdout`
+
+There is no change in the current APi in regard to printing to stdout.  Applications should call `printf` directly as shown below. The console baud rate can be configured via `platform.stdio-baud-rate` in `mbed_app.json`.
```

Comment:
There's currently no direct C++ API to change the console baud rate, aside from that start-up setting, so I think it's just the baud rate.

For comparison other setting `platform.default-serial-baud-rate`, but that is a true default - that's the speed for a `Serial` object where it's never set, and it's perfectly normal to use the real API, rather than that default.

Reply:
Thanks for the clarification.

Reply:
i didnt realise there was `platform.stdio-baud-rate` as well as `platform.default-serial-baud-rate`. I was only aware of the latter.

Reply:
I don't really see the value of `platform.default-serial-baud-rate`. It's a global setting applying to all  serial ports. I struggle to think of a use case - any device with multiple ports is going to want each one (GPS, modem, console) individually set, and code for each of them will be specifying baud rate directly anyway (with settings like `platform.stdio-baud-rate`).

Personally, I would have just had default baud rate be fixed 9600. Just like SPI defaults to 8 bit mode 0, 1MHz - there's no config setting for that. Or indeed like the serial format defaults to 8N1 with no config setting.

I'd be inclined to kill that setting just to reduce clutter and confusion.

Just wondering how to deprecate settings. Maybe change the default to `null`, then have the code do:

    #ifdef MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE
    #define DEFAULT_SERIAL_BAUD_RATE MBED_CONF_PLATFORM_DEFAULT_SERIAL_BAUD_RATE
    #warning "Deprecated blah blah"
    #else
    #define DEFAULT_SERIAL_BAUD_RATE 9600
    #endif